storage, kv: report long-lived iterators

RaduBerinde opened the pull request:
This change enables the new iterator tracking feature in Pebble. Every
5 minutes, we log the iterators that have been open for more than 1
minute.

Because iterator tracking has measurable overhead, we exempt iterators
in hot paths (based on the read category). This required some
read category plumbing fixes and the addition of a new category for
the abort span.

The test utilizes go1.25's `synctest` and will start running when we
upgrade Go. I did run it locally with go 1.25.

Epic: none
Release note: None

jbowens reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

Nice!

> @jbowens reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155021#-:-Ob48pyi7osUn0D2OvR9:bkx9igh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 7 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155021#-:-Ob5Z1sU9vJwshB9hh4k:bg7igkv)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[`pkg/storage/engine.go` line 1446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155021#-Ob5YVBAASp5Ux8NwZ07:-Ob5YVBAASp5Ux8NwZ08:bsfequ7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fddf57d382b1b429c38c3417017d6b2026443f6a/pkg/storage/engine.go#L1446)):*
> ```Go
> 		// Ignore Exclusive and Shared locks. We only care about intents.
> 		MatchMinStr:  lock.Intent,
> 		ReadCategory: fs.IntentResolutionReadCategory,
> ```

how about a comment like
// This is eventually called from the QueryIntent request, so this isn't quite "intent resolution", but we don't want too many categories, and this does relate to intents, so we use this existing category.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155021)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/storage/engine.go` line 1446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155021#-Ob5YVBAASp5Ux8NwZ07:-Ob5pk-U4RIo42talsQn:bexegmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fddf57d382b1b429c38c3417017d6b2026443f6a/pkg/storage/engine.go#L1446)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how about a comment like
// This is eventually called from the QueryIntent request, so this isn't quite "intent resolution", but we don't want too many categories, and this does relate to intents, so we use this existing category.
</blockquote></details>

Done, TFTR!


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
storage, kv: report long-lived iterators

This change enables the new iterator tracking feature in Pebble. Every
5 minutes, we log the iterators that have been open for more than 1
minute.

Because iterator tracking has measurable overhead, we exempt iterators
in hot paths (based on the read category). This required some
read category plumbing fixes and the addition of a new category for
the abort span.

The test utilizes go1.25's `synctest` and will start running when we
upgrade Go. I did run it locally with go 1.25.

Epic: none
Release note: None
